Fix 404 error for GitLab provider

savitaashture opened the pull request:
In GitLab, when there is an MR from a forked repo
to a target repo, it gives a 404 error from
the GetMergeRequestChanges function
This PR addresses the fix

Signed-off-by: Savita Ashture <sashture@redhat.com>

# Changes <!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

<!-- Describe your changes here- ideally you can get that description straight from
your descriptive commit message(s)! -->

# Submitter Checklist

- [ ] 📝 Please ensure your commit message is clear and informative. For guidance on crafting effective commit messages, refer to the How to write a git commit message guide. We prefer the commit message to be included in the PR body itself rather than a link to an external website (ie: Jira ticket).

- [ ] ♽ Before submitting a PR, run make test lint to avoid unnecessary CI processing. For an even more efficient workflow, consider installing [pre-commit](https://pre-commit.com/) and running pre-commit install in the root of this repository.

- [ ] ✨ We use linters to maintain clean and consistent code. Please ensure you've run make lint before submitting a PR. Some linters offer a --fix mode, which can be executed with the command make fix-linters (ensure [markdownlint](https://github.com/DavidAnson/markdownlint) and [golangci-lint](https://github.com/golangci/golangci-lint) tools are installed first).

- [ ] 📖 If you're introducing a user-facing feature or changing existing behavior, please ensure it's properly documented.

- [ ] 🧪 While 100% coverage isn't a requirement, we encourage unit tests for any code changes where possible.

- [ ] 🎁 If feasible, please check if an end-to-end test can be added. See [README](https://github.com/openshift-pipelines/pipelines-as-code/blob/main/test/README.md) for more details.

- [ ] 🔎 If there's any flakiness in the CI tests, don't necessarily ignore it. It's better to address the issue before merging, or provide a valid reason to bypass it if fixing isn't possible (e.g., token rate limitations).

savitaashture committed:
Fix 404 error for GitLab provider

In GitLab, when there is an MR from a forked repo
to a target repo, it gives a 404 error from
the GetMergeRequestChanges function
This PR addresses the fix

Signed-off-by: Savita Ashture <sashture@redhat.com>

savitaashture commented:
cc @gabemontero

gabemontero reviewed (approved):
lgtm

what was decided @savitaashture about the lack of error processing at https://github.com/openshift-pipelines/pipelines-as-code/blob/0e97e1d1afead542c99ce8004a9f800b0998efe8/pkg/provider/gitlab/gitlab.go#L213 ?

chmouel commented:
> lgtm
> 
> what was decided @savitaashture about the lack of error processing at
> 
> https://github.com/openshift-pipelines/pipelines-as-code/blob/0e97e1d1afead542c99ce8004a9f800b0998efe8/pkg/provider/gitlab/gitlab.go#L213

because when you create a Merge Request (MR) from a fork in an upstream repository, the token needs to have write access to the fork repository to create a status. Typically, the token set on the Repository CR doesn't have such broad access, so we can't create a status commit on it.

I explained this poorly in a comment when I first implemented it a few years ago here:

https://github.com/openshift-pipelines/pipelines-as-code/blob/467aebb3d9fbba241cf4ad4001bd9db8ed7a64e9/pkg/provider/gitlab/gitlab.go#L189-L193

I asked our GitLab developer liaison about it but haven't received a response. I'm fairly certain there was a bug report in the GitLab instance addressing this issue.  maybe something has changed since then 🤷🏻

gabemontero commented:
> > lgtm
> > what was decided @savitaashture about the lack of error processing at
> > https://github.com/openshift-pipelines/pipelines-as-code/blob/0e97e1d1afead542c99ce8004a9f800b0998efe8/pkg/provider/gitlab/gitlab.go#L213
> 
> because when you create a Merge Request (MR) from a fork in an upstream repository, the token needs to have write access to the fork repository to create a status. Typically, the token set on the Repository CR doesn't have such broad access, so we can't create a status commit on it.
> 
> I explained this poorly in a comment when I first implemented it a few years ago here:
> 
> https://github.com/openshift-pipelines/pipelines-as-code/blob/467aebb3d9fbba241cf4ad4001bd9db8ed7a64e9/pkg/provider/gitlab/gitlab.go#L189-L193
> 
> I asked our GitLab developer liaison about it but haven't received a response. I'm fairly certain there was a bug report in the GitLab instance addressing this issue. maybe something has changed since then 🤷🏻

thanks for the detailed explanation @chmouel  .... helps immensely 

@savitaashture - in this PR let's get this info in a comment around https://github.com/openshift-pipelines/pipelines-as-code/blob/0e97e1d1afead542c99ce8004a9f800b0998efe8/pkg/provider/gitlab/gitlab.go#L213 so questions around this do not come up again in a few months and we have to re-research the "why"

savitaashture commented:
Thank you @chmouel for explaining reason behind not handling the error

@chmouel @gabemontero I have updated the existing comment 

PTAL Thank you

enarha commented on pkg/provider/gitlab/gitlab.go:
```diff
@@ -198,11 +198,14 @@ func (v *Provider) CreateStatus(_ context.Context, event *info.Event, statusOpts
+	// from a fork in an upstream repository, the token needs to have write access
+	// to the fork repository in order to create a status. However, the token set on the
+	// Repository CR usually doesn't have such broad access, preventing from creating
+	// a status commit on it.
```
What is "status commit"? Do you mean status comment?

gabemontero replied:
actually @savitaashture I think in addition to the clarification here you need to add right before the 
```
//nolint: dogsled
```
comment below that is preceding `_, _, _ = v.Client.Commits.SetCommitStatus(event.SourceProjectID, event.SHA, opt)` an additional comment about ignoring errors because of the write access issues, along with some indication that we cannot reliable discern write access errors from other errors that we should process

chmouel replied:
instead of the error ignore, as mentioned during pac call maybe we could p.eventEmitter in the user namespace an error that we cannot set status commit with the github token due of the err.Error() 

(and then you can remove the nolint dogsleg stuff)

gabemontero replied:
I like the emit event in user namespace idea / removing the nolint

I would still move the comment right above the error processing, but I'll defer to whatever you all want to do

savitaashture replied:
Addressed here : https://github.com/openshift-pipelines/pipelines-as-code/pull/1692/files#r1625635587

savitaashture commented on pkg/provider/gitlab/gitlab.go:
```diff
@@ -198,19 +202,24 @@ func (v *Provider) CreateStatus(_ context.Context, event *info.Event, statusOpts
+	// a status comment on it.
+	// This would work on a push or an MR from a branch within the same repo.
+	// Ignoring errors because of the write access issues,
+	if _, _, err := v.Client.Commits.SetCommitStatus(event.SourceProjectID, event.SHA, opt); err != nil {
+		v.eventEmitter.EmitMessage(v.repo, zap.ErrorLevel, "FailedToSetCommitStatus",
+			"cannot set status with the GitLab token because of: "+err.Error())
+	}
```
@chmouel @gabemontero handled 
PTAL
Now if there are any error it will show in the events of a user namespace
```
$ oc get events -n test-pipelines | grep FailedToSetCommitStatus
<unknown>   Warning   FailedToSetCommitStatus       repository/savitaashture-test                 cannot set status with the GitLab token because of: POST https://gitlab.com/api/v4/projects/35153477/statuses/48de92146745b56fe5e736c8700dd788e4d66223: 404 {message: 404 References for commit Not Found}
<unknown>   Warning   FailedToSetCommitStatus       repository/savitaashture-test                 cannot set status with the GitLab token because of: POST https://gitlab.com/api/v4/projects/57869138/statuses/48de92146745b56fe5e736c8700dd788e4d66223: 403 {message: 403 Forbidden}

```

savitaashture committed:
Fix 404 error for GitLab provider

In GitLab, when there is an MR from a forked repo
to a target repo, it gives a 404 error from
the GetMergeRequestChanges function
This PR addresses the fix

Signed-off-by: Savita Ashture <sashture@redhat.com>

chmouel reviewed (approved):
LGTM, was it tested manually as working ? (since there is no E2E)

gabemontero reviewed (approved):
awesome / LGTM thanks @savitaashture

savitaashture commented:
> LGTM, was it tested manually as working ? (since there is no E2E)

Yes i did test manually 

we need to have a fork repo to test this that's why did not add e2e
